### DISCUSSION TITLE
Top-level await and dynamic imports

### DISCUSSION THREAD

**Discussion by zeen** [TypeStrong/ts-node#1310] [Q&A]
I have a relatively straightforward typescript project, that's importing nodejs internal libraries, some yarn-installed JS modules and other typescript files. I'm trying to add dynamic import with top-level await to a file, and struggling to figure out the right tsconfig.

Example code:
```
import {PassThrough} from 'stream'; // nodejs's built-in stream library
const handlers = [
  await import('A').handler,
  await import('B').handler,
  await import('C').handler,
];
```

**The following two configurations are interesting:**
1. With `"target": "es2019", "module": "commonjs"`:
```
error TS1378: Top-level 'await' expressions are only allowed when the 'module' option is set to 'esnext' or 'system', and the 'target' option is set to 'es2017' or higher.
```
2. With `"target": "es2019", "module": "esnext"`:
```
(node:55) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
...
import { PassThrough } from 'stream';
^^^^^^

SyntaxError: Cannot use import statement outside a module
```

I did add `"type": "module"` to `package.json`, which seems to have no impact.

Interestingly VSCode doesn't show an error in the editor with (2) config above, only ts-node does.

What am I missing? I'm on Node v15, using ts-node 9.1.1, typescript 4.2.4. Target is es2019 simply because that most closely matched features Node v15 actually supported, I can change versions of anything as needed.

**Comment by cspotcode** [comment]
> Interestingly VSCode doesn't show an error in the editor with (2) config above, only ts-node does.

Remember that VSCode is not transforming nor executing the code, only performing static analysis via the language service.

`ts-node` is doing 3x more than VSCode:
- running the compiler's static analysis
- transforming .ts into .js
- asking node to execute the in-memory .js

You probably want to use our ESM loader.  #1007 has instructions.
